### PR TITLE
Add a command-line flag and an environment variable to enable pegen

### LINKED ISSUE #4: Command-line flag to enable and disable the new parser
- [ ] Add a non-functional flag to begin with.
- [ ] Implement the needed functionality, so that enabling the flag has the desired effect.

### LINKED ISSUE #5: Environment variable to enable and disable the new parser
It should be named `PYTHON{SOMETHING}`.

- [ ] Add a non-function env variable to begin with.
- [ ] Actually make the env variable have the desired effect.

### PR DESCRIPTION
**Pull request by lysnikolaou**
This is certainly a WIP and in no way complete. I'm just opening a PR for feedback, for things I might have missed etc.

Closes #4 and closes #5.

### CODE REVIEW

**Comment by lysnikolaou** [comment]
For example, a question that arises is what exactly the difference between `_PyConfig_InitCompatConfig`, `PyConfig_InitPythonConfig` and `PyConfig_InitIsolatedConfig` is.

**Comment by pablogsal** [comment]
> For example, a question that arises is what exactly the difference between `_PyConfig_InitCompatConfig`, `PyConfig_InitPythonConfig` and `PyConfig_InitIsolatedConfig` is.

The first one is private and I think is used because a subset of the config is needed to initialize some parts first but I can ask Victor Stinner for it, for the other two you can read https://docs.python.org/3/c-api/init_config.html#init-python-config and https://docs.python.org/3/c-api/init_config.html#init-isolated-conf

**Comment by lysnikolaou** [comment]
So, at this time invoking python with the `-p` flag indeed enables pegen. But `test_embed` still fails and I can't really find out why. My instinct is that we need to add the new `PyConfig` field to some other function in `initconfig.c` as well. Any other ideas on what's missing?

**Comment by lysnikolaou** [comment]
I dropped Victor an email and he told me what the problem was. Turns out everything is documented here for anyone who might be interested: https://pythondev.readthedocs.io/pyconfig.html.

**Comment by gvanrossum** [comment]
Are you going to update the output of `python -h`?

File: Python/pythonrun.c [hunk 1]
```diff
@@ -1034,9 +1035,12 @@ PyRun_StringFlags(const char *str, int start, PyObject *globals,
 
     // I left this here so when you run 'nm python | grep PyPegen' you can see
     // that the symbol is included and this function is called.
```
Comment (gvanrossum) [inline Python/pythonrun.c:1037]: Time to kill this comment.

**Comment by lysnikolaou** [comment]
> I suppose the environment variable (#5) and the explicit APIs (#6) will be separate PRs?

Since adding the environment variable is such a small change, I did the work on this PR.

**Comment by lysnikolaou** [comment]
In `PyRun_InteractiveOneObjectEx` there is the "problem" that the filename string is not present. Would we prefer to add another argument `filename_str` to that function or create a new pegen C API function called `PyPegen_ASTFromFilePointer` or something similar?

**Comment by pablogsal** [comment]
We should run the test suite with the new parser or at least some selected files like `test_ast` and `test_grammar`.

**Comment by pablogsal** [comment]
> We should run the test suite with the new parser or at least some selected files like `test_ast`, `test_grammar`, ... to know that it works.

We can do this maybe by exporting the environment variable in the test.

**Comment by pablogsal** [comment]
> When we switch to making the new parser the default, would we just invert the meaning of the `-p` flag and the `PYTHONPEG` env var? That seems weird, because it would mean there'd be no safe way for early adopters to force the PEG parser on (or off). It would be nicer if we could instead have the flag syntax be `-p old` and `-p new`, where we'd just switch the default. For the env var maybe we could use `PYTHONPARSER=old` and `PYTHONPARSER=new`. Other flag values would be errors; other env var values would be ignored.

I think that makes sense. There are similar environment variables already like `PYTHONMALLOC=debug` or `PYTHONMALLOC=malloc`. Another possibility is just to design the flag so it only allows to call the old parser and just start with the new by default.

**Comment by gvanrossum** [comment]
> Another possibility is just to design the flag so it only allows to call the old parser and just start with the new by default.

Hm, maybe we've been acting with too much caution. We'd have to change this in the PEP. @lysnikolaou what do you think?

**Comment by lysnikolaou** [comment]
> > Another possibility is just to design the flag so it only allows to call the old parser and just start with the new by default.
> 
> Hm, maybe we've been acting with too much caution. We'd have to change this in the PEP. @lysnikolaou what do you think?

This is definitely true. There's only around 26 days between alpha6 and beta1. Many people won't really know about the change, let alone use and test it in these first 26 days.

Then again, having something `-p new` and `PYTHONPARSER=new` doesn't seem like that much extra work and it gives us a bit more flexibility, if you're reluctant about having pegen be the default.

I'm certainly leaning on making pegen the default parser. I don't think it makes much of a difference. Also, it would certainly give me much more confidence, because it would force that the whole CPython test suite passes with pegen enabled (which it still doesn't) really early on, which I think is a good thing.

**Comment by gvanrossum** [comment]
Let's implement the new/old option. That also gives us a negotiating position -- if people think it's too soon for Python 3.9, we can propose to have it available but off by default in 3.9, and reconsider in 3.10.

Also it seems that if the flag was on by default the test suite doesn't pass yet? We would need to fix that before turning it on by default (and the fix oughtn't be "run the test suite with the old parser" :-). That would be a separate PR I presume.

**Comment by lysnikolaou** [comment]
> Also it seems that if the flag was on by default the test suite doesn't pass yet? We would need to fix that before turning it on by default (and the fix oughtn't be "run the test suite with the old parser" :-).

It's going to be multiple PRs most probably. Starting with #32.

**Comment by lysnikolaou** [comment]
> In `PyRun_InteractiveOneObjectEx` there is the "problem" that the filename string is not present. Would we prefer to add another argument `filename_str` to that function or create a new pegen C API function called `PyPegen_ASTFromFilePointer` or something similar?

This is still an open question.

File: Python/initconfig.c [hunk 2]
```diff
@@ -2003,6 +2017,14 @@ config_parse_cmdline(PyConfig *config, PyWideStringList *warnoptions,
             config->use_hash_seed = 0;
             break;
 
+        case 'p':
+            if (wcscmp(_PyOS_optarg, L" new") == 0) {
```
Comment (lysnikolaou) [inline Python/initconfig.c:2021]: Another question that I haven't found the answer to is why we need the leading space in the comparison here. It doesn't seem to work this way for the other command-line options.

File: Doc/using/cmdline.rst [hunk 3]
```diff
@@ -252,10 +252,10 @@ Miscellaneous options
    options).  See also :envvar:`PYTHONDEBUG`.
 
 
-.. cmdoption:: -p
+.. cmdoption:: -p old|new
 
-   Turn on the new PEG parser and disable the current one.
-   See also :envvar:`PYTHONPEG`.
+   Use the new PEG parser, if arg is ``new``.  If arg is ``old``, use the old
+   parser.  See also :envvar:`PYTHONPEG`.
```
Comment (gvanrossum) [inline Doc/using/cmdline.rst:258]:
```suggestion
   parser.  See also :envvar:`PYTHONPARSER`.
```

File: Python/initconfig.c (hunk 2 above)
Comment (gvanrossum) [reply Python/initconfig.c:2021]: Are you sure? For me it only works if I remove the leading space.

Perhaps your test is somehow testing it with the single argument `"-p new"` ? There are two forms that should work: a single argument `"-pnew"` or two arguments `"-p" "new"`.

FWIW the only way I've found to verify which parser is being used is to depend on a known bug (#31):
```
(v39) ~/cpygen$ ./python.exe -c 'f(x=1,x=1)'
  File "<string>", line 1
SyntaxError: keyword argument repeated
(v39) ~/cpygen$ ./python.exe -pnew -c 'f(x=1,x=1)'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
NameError: name 'f' is not defined
(v39) ~/cpygen$ 
```

This was with my fix; without the fix, I'd have to use `"-p new"`.
```suggestion
            if (wcscmp(_PyOS_optarg, L"new") == 0) {
```

File: Python/initconfig.c [hunk 4]
```diff
@@ -2009,7 +2018,11 @@ config_parse_cmdline(PyConfig *config, PyWideStringList *warnoptions,
             break;
 
         case 'p':
-            config->use_peg = 1;
+            if (wcscmp(_PyOS_optarg, L" new") == 0) {
+                config->use_peg = 1;
+            } else if (wcscmp(_PyOS_optarg, L" old") == 0) {
```
Comment (gvanrossum) [inline Python/initconfig.c:2023]:
```suggestion
            } else if (wcscmp(_PyOS_optarg, L"old") == 0) {
```

File: Programs/_testembed.c [hunk 5]
```diff
@@ -486,6 +486,9 @@ static int test_init_from_config(void)
 
     config.install_signal_handlers = 0;
 
+    putenv("PYTHONPEG=");
```
Comment (gvanrossum) [inline Programs/_testembed.c:489]:
```suggestion
    putenv("PYTHONPARSER=");
```

File: Python/initconfig.c [hunk 6]
```diff
@@ -1319,6 +1324,15 @@ config_read_env_vars(PyConfig *config)
         config->malloc_stats = 1;
     }
 
+    const char *parser = config_get_env(config, "PYTHONPARSER");
+    if (parser) {
+        if (strcmp(parser, "new") == 0) {
+            config->use_peg = 1;
+        } else if (strcmp(parser, "old") == 0) {
```
Comment (gvanrossum) [inline Python/initconfig.c:1331]: PEP 7
```suggestion
        }
        else if (strcmp(parser, "old") == 0) {
```

**Comment by gvanrossum** [comment]
I just discovered that interactive mode doesn't use the new parser yet even if you specify `-pnew`.

(Also, I noticed a few more PEP 7 violations -- basically you cannot have `} else`.)

**Comment by lysnikolaou** [comment]
> I just discovered that interactive mode doesn't use the new parser yet even if you specify `-pnew`.

This is related to [my comment above](https://github.com/we-like-parsers/cpython/pull/29#issuecomment-607194384).

**Comment by gvanrossum** [comment]
> This is related to my comment above.

How? Sorry for being dense.

**Comment by lysnikolaou** [comment]
> > This is related to my comment above.
> 
> How? Sorry for being dense.

In order for interactive mode to use pegen, we need to have the same `if` check in `PyRun_InteractiveOneObjectEx`. To do so, we first need to solve the problem, that we don't the filename string in that function. There is only a `FILE *` and a `PyObject *`. 

We can either implement a new C API function that accepts a `FILE *` or we can just call `PyUnicode_AsUTF8` with the filename object. What do you think is best?

**Comment by gvanrossum** [comment]
> We can either implement a new C API function that accepts a `FILE *` or we can just call `PyUnicode_AsUTF8` with the filename object. What do you think is best?

We need to make a function taking a `FILE *`, because in this case the file is already open. Typically it's the UNIX process stdin, which may not have a filename (e.g. a pipe).

**Pull request merged**